src/Functions/array/mapFill.cpp

```diff
@@ -0,0 +1,308 @@
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnTuple.h>
+#include <Columns/ColumnVector.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include "Core/ColumnWithTypeAndName.h"
+#include "DataTypes/IDataType.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+class FunctionMapFill : public IFunction
+{
+public:
+    static constexpr auto name = "mapFill";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionMapFill>(); }
+
+private:
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() < 2)
+            throw Exception{getName() + " accepts at least two arrays for key and value", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH};
+
+        if (arguments.size() > 3)
+            throw Exception{"too many arguments in " + getName() + " call", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH};
+
+        const DataTypeArray * key_array_type = checkAndGetDataType<DataTypeArray>(arguments[0].get());
+        const DataTypeArray * val_array_type = checkAndGetDataType<DataTypeArray>(arguments[1].get());
+
+        if (!key_array_type || !val_array_type)
+            throw Exception{getName() + " accepts two arrays for key and value", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+
+        DataTypePtr keys_type = key_array_type->getNestedType();
+        WhichDataType which_key(keys_type);
+        if (!(which_key.isNativeInt() || which_key.isNativeUInt()))
+        {
+            throw Exception(
+                "Keys for " + getName() + " should be of native integer type (signed or unsigned)", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+
+        if (arguments.size() == 3)
+        {
+            DataTypePtr max_key_type = arguments[2];
+            WhichDataType which_max_key(max_key_type);
+
+            if (which_max_key.isNullable())
+                throw Exception(
+                    "Max key argument in arguments of function " + getName() + " can not be Nullable",
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+            if (keys_type->getTypeId() != max_key_type->getTypeId())
```

Comment:
The idea was to just provide correct limits for `max` key. For example if we have `key` as `Array(UInt8)` and `max` greater than 255, we're basically have two options: to cut off `max` or change the returning type of `key` to UInt16. I wanted to avoid that.